Source/Core/Core/AchievementManager.cpp

```diff
@@ -1020,8 +1037,15 @@ void AchievementManager::FetchBadge(AchievementManager::Badge* badge, u32 badge_
 
     if (!LoadPNGTexture(badge, *http_response))
     {
-      ERROR_LOG_FMT(ACHIEVEMENTS, "Default game badge '{}' failed to load",
-                    DEFAULT_GAME_BADGE_FILENAME);
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Badge '{}' failed to load", name_to_fetch);
+    }
+
+    std::string cache_base = File::GetUserPath(D_RETROACHIEVEMENTSCACHE_IDX);
+    std::string cache_path = fmt::format("{}/{}-{}.png", cache_base, badge_type, name_to_fetch);
+    if (!Common::SavePNG(cache_path, badge->data.data(), Common::ImageByteFormat::RGBA,
+                         badge->width, badge->height, badge->width * 4))
```

Comment:
Two things here:

- This should be two-step. Write to a temp file, check if the write succeeded, if yes rename the temp file to the target file, if no delete the temp file.
- Encoding `name_to_fetch` into the filename here is... tricky, since it comes from a server and could contain anything, such characters that are not valid in the host file system, or even a path traversal attack from a name containing slashes. Originally I was going to suggest escaping the path (`Common::EscapeFileName()` from `NandPaths.h`), but I think it's probably better to just hash it with SHA1 or so and then use the hash. The chance of collision is negligible.